File: 03-march/13-minimum-number-of-seconds-to-make-mountain-height-zero/solution.java
class Solution {
    public long minNumberOfSeconds(int mountainHeight, int[] workerTimes) {
        int fastest = Integer.MAX_VALUE;
        for (int workerTime : workerTimes) {
            fastest = Math.min(fastest, workerTime);
        }

        long lo = 0;
        long hi = (long) fastest * mountainHeight * (mountainHeight + 1) / 2;

        while (lo < hi) {
            long mid = lo + (hi - lo) / 2;
            if (feasible(mid, mountainHeight, workerTimes)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }

        return lo;
    }

    private boolean feasible(long timeLimit, int mountainHeight, int[] workerTimes) {
        long removed = 0;
        for (int workerTime : workerTimes) {
            removed += maxUnits(timeLimit, workerTime, mountainHeight);
            if (removed >= mountainHeight) {
                return true;
            }
        }
        return false;
    }

    private int maxUnits(long timeLimit, int workerTime, int mountainHeight) {
        int lo = 0;
        int hi = mountainHeight;

        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;
            long need = (long) workerTime * mid * (mid + 1) / 2;
            if (need <= timeLimit) {
                lo = mid;
            } else {
                hi = mid - 1;
            }
        }

        return lo;
    }
}

File: 03-march/13-minimum-number-of-seconds-to-make-mountain-height-zero/solution.py
class Solution:
    def minNumberOfSeconds(self, mountainHeight: int, workerTimes: list[int]) -> int:
        def max_units(time_limit: int, worker_time: int) -> int:
            lo = 0
            hi = mountainHeight
            while lo < hi:
                mid = (lo + hi + 1) // 2
                if worker_time * mid * (mid + 1) // 2 <= time_limit:
                    lo = mid
                else:
                    hi = mid - 1
            return lo

        def feasible(time_limit: int) -> bool:
            removed = 0
            for worker_time in workerTimes:
                removed += max_units(time_limit, worker_time)
                if removed >= mountainHeight:
                    return True
            return False

        fastest = min(workerTimes)
        lo = 0
        hi = fastest * mountainHeight * (mountainHeight + 1) // 2

        while lo < hi:
            mid = (lo + hi) // 2
            if feasible(mid):
                hi = mid
            else:
                lo = mid + 1

        return lo
